tests/nn/data_parallel/test_fsdp_regnet.py

```diff
@@ -343,8 +342,36 @@ def test1(temp_files, ddp_ref, precision, flatten):
     fsdp_config["mixed_precision"] = precision == "mixed"
     fsdp_config["flatten_parameters"] = flatten == "flatten"
 
-    if fsdp_config["mixed_precision"] and torch_cuda_version() < (11, 0):
-        pytest.skip("Only CUDA 11 is supported with AMP equivalency")
+    if fsdp_config["mixed_precision"]:
+        # For convolution layers, we need fp32_compute_dtype. Otherwise, we run
+        # into conv layer errors like below with mixed precision with larger
+        # regnet128Gf models when flatten_parameter is true:
+        #
+        #   RuntimeError: Unable to find a valid cuDNN algorithm to run convolution
+        #   or
+        #   RuntimeError: cuDNN error: CUDNN_STATUS_BAD_PARAM
+        #
+        # If you know the solutions to the above issue, please let me (min-xu-ai) know. Thanks!
+        #
+        # Also, without fp32_compute_dtype, Linear layer with bias=True won't match DDP
+        # numerically, which may not be necessary for convergence depending on the model and
+        # the optimizer and a number of other factors.
```

Comment:
Can you clarify what this is referring to, perhaps with a small repro?

For some context: the recommended mixed precision recipe for PyTorch is to use an FP32 model and autocast weights to FP16 on-the-fly.

Contrast that with `FSDP(mixed_precision=True)`, where FSDP will move everything to FP16 internally, and will rely on autocast to up-cast things to FP32 when needed (e.g., for Softmax).

When comparing these two recipes, they are not expected to give identical results. Is there some specific concern for Linear with bias, or is this just a general observation about differences between the two mixed precision recipes?

Reply:
I will actually preserve this part in this test. Will be added in a new commit. So the repro will be in this test.

Here is what I think is happening:

without bias: the operation is sum(x * y for x, y in zip(vector1, vector2)) for each resulting weight.
with bias: the operation is sum(x*y for x, y in zip(vector1, vector2)) + bias for each weight.

The vector1/vector2 are in fp16 whether or not they were originally in fp32 or fp16, because of auto-cast.

However, with bias, pytorch's mode will be adding a fp32 term but in FSDP, we will be adding a fp16 term. I suspect that it is the issue causing the difference. What do you think?

Reply:
> However, with bias, pytorch's mode will be adding a fp32 term but in FSDP, we will be adding a fp16 term.

Are you sure? That could be possible, but I'm not able to repro locally (PyTorch 1.8.1):
```python
>>> import torch
>>>
>>> m = torch.nn.Linear(1024, 4096, bias=True).cuda()
>>> m2 = torch.nn.Linear(1024, 4096, bias=True).cuda()
>>> m2.load_state_dict(m.state_dict())
>>>
>>> m2 = m2.half()
>>> m.bias.dtype
torch.float32
>>> m2.bias.dtype
torch.float16
>>>
>>> with torch.cuda.amp.autocast():
...   y = m(x)
...   y2 = m2(x)
...
>>> y.dtype
torch.float16
>>> y2.dtype
torch.float16
>>> torch.testing.assert_allclose(y.float(), y2.float())
```

Reply:
interesting, with my 1.6:
```
  File "/private/home/m1n/e/py38_fs/lib/python3.8/site-packages/torch/testing/__init__.py", line 223, in assert_allclose
    raise AssertionError(msg)
AssertionError: With rtol=0.0001 and atol=1e-05, found 202 element(s) (out of 4096) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 0.00048828125 (0.56494140625 vs. 0.564453125), which occurred at index 1915.
```

same code as yours.

Reply:
😮